x-pack/plugins/uptime/server/lib/requests/get_monitor_availability.ts

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { UMElasticsearchQueryFn } from '../adapters';
+import { GetMonitorAvailabilityParams } from '../../../common/runtime_types';
+
+export interface AvailabilityKey {
+  monitorId: string;
+  location: string;
+  name: string;
+  url: string;
+}
+
+export interface GetMonitorAvailabilityResult {
+  monitorId: string;
+  location: string;
+  name: string;
+  url: string;
+  up: number;
+  down: number;
+  availabilityRatio: number | null;
+}
+
+export const formatBuckets = async (buckets: any[]): Promise<GetMonitorAvailabilityResult[]> => {
+  return buckets.map(({ key, up_sum, down_sum, ratio }: any) => ({
+    ...key,
+    up: up_sum.value,
+    down: down_sum.value,
+    availabilityRatio: ratio.value,
+  }));
+};
+
+export const getMonitorAvailability: UMElasticsearchQueryFn<
+  GetMonitorAvailabilityParams,
+  GetMonitorAvailabilityResult[]
+> = async ({ callES, dynamicSettings, range, rangeUnit, threshold: thresholdString, filters }) => {
+  const queryResults: Array<Promise<GetMonitorAvailabilityResult[]>> = [];
+  let afterKey: AvailabilityKey | undefined;
+  const gte = `now-${range}${rangeUnit}`;
+  const threshold = Number(thresholdString) / 100;
+  if (threshold <= 0 || threshold > 1.0) {
+    throw new Error(
+      `Invalid availability threshold value ${thresholdString}. The value must be between 0 and 100`
+    );
+  }
+
+  do {
+    const esParams: any = {
+      index: dynamicSettings.heartbeatIndices,
+      body: {
+        query: {
+          bool: {
+            filter: [
+              {
+                range: {
+                  '@timestamp': {
+                    gte,
+                    lte: 'now',
+                  },
+                },
+              },
+            ],
+          },
+        },
+        size: 0,
+        aggs: {
+          monitors: {
+            composite: {
+              size: 2000,
+              sources: [
+                {
+                  monitorId: {
+                    terms: {
+                      field: 'monitor.id',
+                    },
+                  },
+                },
+                {
+                  location: {
+                    terms: {
+                      field: 'observer.geo.name',
+                      missing_bucket: true,
+                    },
+                  },
+                },
+                {
+                  name: {
```

Comment:
If we aggregate by name and URL that means that users can't change these values without impacting the calculation of availability, thus losing history. We could deduce these fields via top_hits instead, getting the most recent doc for each bucket.

Reply:
That thought had crossed my mind too, I agree we should just add a top_hits agg.

Reply:
I've added a top_hits to retrieve the most recent values for name and URL per ID/location. See https://github.com/elastic/kibana/pull/70284/commits/178f1ad0ca3f47fac223e93e5cfb57f680421e87 and https://github.com/elastic/kibana/pull/70284/commits/f6c33849b9e31d4c029dfcd36ad8a93ac93a43e5.